performance improvements to Flag access

dskells opened the pull request:
Removes outer class access and other operations if not needed

 - Use bulk flag query where possible
 - Push down auto-initialization logic into overrides in runtime reflection
 - Avoid lookup of the phase-indexed flag mask for flags known to be phase independent (e.g. `ERROR`)
 - Make more flag masks constant foldable

dskells committed:
remove duplicated code

dskells commented:
```
after 0 best 50%
                 Run Name    Cycle    samples               Wall time (ms)           All Wall time (ms)                      CPU(ms)               Idle time (ms)               Allocated(MBs)
               before-dan       1      50     6,612.06 [-4.22% +4.52%]     6,612.06 [-4.22% +4.52%]     6,518.32 [-3.89% +4.75%]           0.00 [ NaN%  NaN%]     1,281.31 [-2.39% +3.88%]
               before-dan       2      50     6,525.90 [-3.16% +4.12%]     6,525.90 [-3.16% +4.12%]     6,424.22 [-3.21% +3.84%]           0.00 [ NaN%  NaN%]     1,263.62 [-2.29% +3.86%]
               before-dan       3      50     6,590.64 [-4.46% +4.18%]     6,590.64 [-4.46% +4.18%]     6,496.12 [-5.24% +4.62%]           0.00 [ NaN%  NaN%]     1,277.04 [-2.39% +3.84%]
                after-dan       1      50     6,392.98 [-4.40% +4.54%]     6,392.98 [-4.40% +4.54%]     6,291.76 [-3.65% +4.04%]           0.00 [ NaN%  NaN%]     1,251.92 [-2.36% +3.86%]
                after-dan       2      50     6,328.84 [-3.35% +3.00%]     6,328.84 [-3.35% +3.00%]     6,288.26 [-3.60% +3.37%]           0.00 [ NaN%  NaN%]     1,252.19 [-2.44% +3.95%]
                after-dan       3      50     6,398.92 [-3.20% +4.11%]     6,398.92 [-3.20% +4.11%]     6,304.66 [-2.85% +4.34%]           0.00 [ NaN%  NaN%]     1,260.49 [-2.42% +3.87%]
-----
after 0 best 50%, phase jvm
                 Run Name    Cycle    samples               Wall time (ms)           All Wall time (ms)                      CPU(ms)               Idle time (ms)               Allocated(MBs)
               before-dan       1      50     1,613.90 [-2.22% +2.55%]     1,613.90 [-2.22% +2.55%]     1,601.20 [-2.45% +1.49%]           0.00 [ NaN%  NaN%]       318.75 [-1.62% +1.97%]
               before-dan       2      50     1,587.86 [-1.82% +1.96%]     1,587.86 [-1.82% +1.96%]     1,576.48 [-1.93% +2.06%]           0.00 [ NaN%  NaN%]       315.96 [-1.05% +1.46%]
               before-dan       3      50     1,619.24 [-3.41% +1.71%]     1,619.24 [-3.41% +1.71%]     1,595.88 [-3.13% +1.82%]           0.00 [ NaN%  NaN%]       317.83 [-1.57% +1.76%]
                after-dan       1      50     1,511.82 [-4.62% +3.25%]     1,511.82 [-4.62% +3.25%]     1,474.64 [-2.55% +1.72%]           0.00 [ NaN%  NaN%]       314.45 [-1.36% +1.82%]
                after-dan       2      50     1,474.32 [-2.87% +1.74%]     1,474.32 [-2.87% +1.74%]     1,467.66 [-2.09% +1.11%]           0.00 [ NaN%  NaN%]       314.26 [-1.77% +2.28%]
                after-dan       3      50     1,480.30 [-2.38% +2.68%]     1,480.30 [-2.38% +2.68%]     1,469.70 [-2.22% +2.06%]           0.00 [ NaN%  NaN%]       322.45 [-1.54% +1.92%]
-----
after 0 best 50%, phase typer
                 Run Name    Cycle    samples               Wall time (ms)           All Wall time (ms)                      CPU(ms)               Idle time (ms)               Allocated(MBs)
               before-dan       1      50     1,828.16 [-1.60% +1.96%]     1,828.16 [-1.60% +1.96%]     1,827.36 [-1.72% +1.73%]           0.00 [ NaN%  NaN%]       356.44 [-0.01% +0.01%]
               before-dan       2      50     1,834.38 [-1.49% +1.89%]     1,834.38 [-1.49% +1.89%]     1,835.52 [-1.28% +2.15%]           0.00 [ NaN%  NaN%]       354.56 [-0.01% +0.01%]
               before-dan       3      50     1,818.74 [-1.64% +1.99%]     1,818.74 [-1.64% +1.99%]     1,818.86 [-2.08% +2.21%]           0.00 [ NaN%  NaN%]       356.14 [-0.01% +0.01%]
                after-dan       1      50     1,793.22 [-1.85% +2.39%]     1,793.22 [-1.85% +2.39%]     1,793.54 [-2.43% +2.76%]           0.00 [ NaN%  NaN%]       345.77 [-0.01% +0.01%]
                after-dan       2      50     1,768.58 [-1.67% +1.89%]     1,768.58 [-1.67% +1.89%]     1,769.28 [-1.99% +1.51%]           0.00 [ NaN%  NaN%]       346.81 [-0.01% +0.01%]
                after-dan       3      50     1,803.98 [-1.27% +1.50%]     1,803.98 [-1.27% +1.50%]     1,803.20 [-2.12% +1.38%]           0.00 [ NaN%  NaN%]       346.92 [-0.01% +0.01%]
-----

```

retronym commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -733,27 +734,27 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+        f = flags
+      }
+      f & mask
     }
```
An alternative design would be to make `getFlag` non final, and override it in [`SynchronizedSymbol`](https://github.com/scala/scala/blob/e0967443a8fff62bb8a8ae193b4cb3f541a86fe4/src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala#L34). Assuming that runtime reflection is not used in same classloader that is running the batch compiler, the JVM would treat this method as effectively final and inline it if needed.

We didn't have great performance testing infrastructure when we tried to accommodate the needs of runtime reflection into the compiler implementation, but now that we have better benchmarks, we can revisit these sort of implementation choices.

retronym commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -733,27 +734,27 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     final def getFlag(mask: Long): Long = {
-      if (!isCompilerUniverse && !isThreadsafe(purpose = FlagOps(mask))) initialize
-      flags & mask
+      var f = if ((mask & PhaseIndependentFlags) == mask) rawflags else flags
```
👍 for this part, I tried something [similar in a WIP branch](https://github.com/scala/scala/compare/2.13.x...retronym:faster/round3#diff-6cc67f90cb479803e84e8068d49663fbR816) by special casing `isError`, but this looks cleaner.

retronym commented on src/reflect/scala/reflect/internal/Flags.scala:
```diff
@@ -195,18 +195,21 @@ class Flags extends ModifierFlags {
   final val AntiShift     = 56
+  final val InCompilerUniverse = 0x8000000000000000L
+
+  final val PhaseIndependentFlags = 0x8007FFFFFFFFFE08L //(InitialFlags & ~(LateFlags >>> LateShift) & ~(AntiFlags >>> AntiShift)) | InCompilerUniverse
```
I'd prefer to have the comment as code, assuming you can get the constant folder to play along, as I did in #6445. An alternative would be to add an `assert(0x8007FFFFFFFFFE08L == .... `) below.

retronym replied:
`InCompilerUniverse` can be removed if the overriding idea works out.

retronym commented:
The build for one of the commits got stuck with:

```
"pool-2-thread-6" #16 prio=5 os_prio=0 tid=0x00007f9868256800 nid=0x2863 runnable [0x00007f986c5f3000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.SetLike.isEmpty(SetLike.scala:164)
	at scala.collection.SetLike.isEmpty$(SetLike.scala:164)
	at scala.collection.AbstractSet.isEmpty(Set.scala:47)
	at scala.tools.nsc.transform.patmat.Solving$Solver.$anonfun$findTseitinModel0$1(Solving.scala:521)
	at scala.tools.nsc.transform.patmat.Solving$Solver.$anonfun$findTseitinModel0$1$adapted(Solving.scala:521)
	at scala.tools.nsc.transform.patmat.Solving$Solver$$Lambda$2424/803027833.apply(Unknown Source)
	at scala.collection.IndexedSeqOptimized.prefixLengthImpl(IndexedSeqOptimized.scala:37)
	at scala.collection.IndexedSeqOptimized.exists(IndexedSeqOptimized.scala:45)
	at scala.collection.IndexedSeqOptimized.exists$(IndexedSeqOptimized.scala:45)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findTseitinModel0(Solving.scala:521)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findTseitinModelFor(Solving.scala:475)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findTseitinModelFor$(Solving.scala:470)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.findTseitinModelFor(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findAllModels$1(Solving.scala:427)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findAllModelsFor(Solving.scala:442)
	at scala.tools.nsc.transform.patmat.Solving$Solver.findAllModelsFor$(Solving.scala:398)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.findAllModelsFor(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.findAllModelsFor(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.MatchAnalysis$MatchAnalyzer.exhaustive(MatchAnalysis.scala:558)
	at scala.tools.nsc.transform.patmat.MatchAnalysis$MatchAnalyzer.exhaustive$(MatchAnalysis.scala:513)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.exhaustive(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.MatchAnalysis$MatchAnalyzer.analyzeCases(MatchAnalysis.scala:916)
	at scala.tools.nsc.transform.patmat.MatchAnalysis$MatchAnalyzer.analyzeCases$(MatchAnalysis.scala:909)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.analyzeCases(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.MatchTreeMaking$TreeMakers.$anonfun$combineCasesNoSubstOnly$6(MatchTreeMaking.scala:610)
	at scala.tools.nsc.transform.patmat.MatchTreeMaking$TreeMakers.combineCasesNoSubstOnly(MatchTreeMaking.scala:593)
	at scala.tools.nsc.transform.patmat.MatchTreeMaking$TreeMakers.combineCasesNoSubstOnly$(MatchTreeMaking.scala:556)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.combineCasesNoSubstOnly(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.MatchTreeMaking$TreeMakers.combineCases(MatchTreeMaking.scala:552)
	at scala.tools.nsc.transform.patmat.MatchTreeMaking$TreeMakers.combineCases$(MatchTreeMaking.scala:549)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.combineCases(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.MatchTranslation$MatchTranslator.translateMatch(MatchTranslation.scala:227)
	at scala.tools.nsc.transform.patmat.MatchTranslation$MatchTranslator.translateMatch$(MatchTranslation.scala:195)
	at scala.tools.nsc.transform.patmat.PatternMatching$OptimizingMatchTranslator.translateMatch(PatternMatching.scala:89)
	at scala.tools.nsc.transform.patmat.PatternMatching$MatchTransformer.transform(PatternMatching.scala:62)
	at scala.tools.nsc.transform.patmat.PatternMatching$MatchTransformer.transform(PatternMatching.scala:57)
```

That problem is unrelated to this PR.

retronym commented:
/rebuild

dskells committed:
avoid unneeded operations on some Flag access,
e.g. outer reference for phase and some bit twiddling
adjusted mima filters for additional methods

dskells committed:
use inheritance to avoid initialisation checks for privateWithin
avoid repeated calls to privateWithin from the same method
adjust mima filters

mkeskells commented:
/rebuild

mkeskells commented:
@retronym how do I see from the MIMA report what the binary compatibility is (presume its the final vals, but would like to see the diagnosis

mkeskells commented:
seems to be around 1.5 - 2% from  my tests - not worth publishing details until we resolve the MIMA issues though

retronym commented:
The MiMa output is in the full console [log](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main/494/consoleFull) of the failed build:

```
[test] [error]  * abstract synthetic method scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$getFlag(Long)Long in interface scala.reflect.runtime.SynchronizedSymbols#SynchronizedSymbol is present only in current version
[test] [error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.reflect.runtime.SynchronizedSymbols#SynchronizedSymbol.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$getFlag")
[test] [error]  * method getFlag(Long)Long in interface scala.reflect.runtime.SynchronizedSymbols#SynchronizedSymbol does not have a correspondent in other version
[test] [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.SynchronizedSymbols#SynchronizedSymbol.getFlag")
```

The whitelists are in `src/*/mima-filters/*`

mkeskells commented:
@retronym thanks - would there changes be acceptable to put in the mima-filters. Would have thought that the Flags and Symbols are safe to change, but I don't know the policy around this

Alternatively @dskells could defer the val -> final val to 2.13, and limit the visibility of the new val to [scala]?

dskells committed:
use inheritance to avoid initialisation checks
avoid varags in helper methods to reduce memory pressure
make annotation helper functions final where appropriate
adjust mima filters

dskells committed:
reduce number of getFlag and related calls,
combine call to check multiple flags in a single call where applicable

retronym commented:
What's the status of this PR? Ready to go from your end? Its looking pretty good to me.

retronym commented:
I'm happy for you to manually do the constant folding if needed, and add a documentary assert below.

dskells committed:
add an assertion for PhaseIndependentFlags

mkeskells commented:
/rebuild

retronym commented:
@mkeskells Rebasing on 2.13.x, to get #6548, would fix that test failure.

retronym reviewed (approved):
Great work, @dskells.
